src/backend/distributed/executor/citus_custom_scan.c

```diff
@@ -245,20 +245,33 @@ CitusBeginScanWithCoordinatorProcessing(CustomScanState *node, EState *estate, i
 
 	/* citus only evaluates functions for modification queries */
 	bool modifyQueryRequiresMasterEvaluation =
-		workerJob->requiresMasterEvaluation && jobQuery->commandType != CMD_SELECT;
+		jobQuery->commandType != CMD_SELECT &&
+		(workerJob->requiresMasterEvaluation || workerJob->deferredPruning);
 
 	/*
 	 * ExecuteMasterEvaluableFunctions handles both function evalation
 	 * and parameter evaluation. Pruning is most likely deferred because
 	 * there is a parameter on the distribution key. So, evaluate in both
 	 * cases.
 	 */
-	bool shoudEvaluteFunctionsOrParams =
-		modifyQueryRequiresMasterEvaluation || workerJob->deferredPruning;
-	if (shoudEvaluteFunctionsOrParams)
+	if (modifyQueryRequiresMasterEvaluation)
```

Comment:
with #3388, I tried to make sure that we don't have any specific logic for modify  vs SELECT, but it turns out that my push broke some cases. So,  the aim of  these changes it  make it back as it  used to be. For modifications, to the function/param evaluation, for select queries only param evaluation when needed, otherwise do nothing.